### PR TITLE
AT-5882 View a user’s portfolio permissions

### PR DESCRIPTION
**Pull request by jGonzalez-CCPO**
Description
As a CCPO admin, I want to see all the portfolios a user is part of, so that I can quickly find any portfolios that the user asks for my help with.
Out of scope for this ticket: the … button that leads to editing permissions. That is covered in AT-5883: Drop down menu to edit portfolio permissionsTO DO 
See below for an image of the component and see the User Detail View v2 page in Abstract for design specs. 
Note: Abstract/Invision links will only be available until June 2021. A PNG version of the view from the above Abstract link is retained for reference on Confluence here: User Detail View v2.png. Please see Design Files and Assets  in Confluence for instructions on how you can view design specs using Sketch.

Acceptance Criteria
the table should include all portfolios for which the selected user has a role
the table should include:
portfolio name
user’s role in the portfolio
portfolio name should be sortable alphabetically
default sort is portfolio in alphabetical order
if the CCPO admin clicks the portfolio, they’re brought to the portfolio detail view

### CODE REVIEW

**Comment by jGonzalez-CCPO** [comment]
Need to finish coverage

**Comment by jGonzalez-CCPO** [comment]
Ready for PR

File: atat/routes/admin/admin.py [hunk 1]
```diff
@@ -112,5 +115,26 @@ def all_users_page():
...
+    next_ = request.args.get("next")
+
+    return render_template(
+        "admin/user.html",
+        next=next_,
+        form=form,
+        user=user,
+        portfolio_roles=portfolio_roles_sort,
+        order_asc=order_asc,
+        mindate=pendulum.now(tz="UTC").subtract(days=365),
```
Comment (kylelaker-ccpo) [inline atat/routes/admin/admin.py:138]:
```suggestion
        mindate=pendulum.now(tz="UTC").subtract(years=1),
```

File: atat/routes/admin/admin.py [hunk 2]
```diff
@@ -112,5 +115,26 @@ def all_users_page():
...
+    order_asc = request.args.get("order-asc", "1")
     user = Users.get(user_id)
-    return render_template("admin/user.html", user=user)
+    portfolio_roles = PortfolioRoles.get_all_by_user(user_id)
+    portfolio_roles_sort = sorted(
+        portfolio_roles,
+        key=lambda portfolio_role: portfolio_role.PortfolioRole.portfolio.name,
+    )
+    if order_asc == "-1":
+        portfolio_roles_sort = portfolio_roles_sort[::-1]
```
Comment (kylelaker-ccpo) [inline atat/routes/admin/admin.py:126]:
```suggestion
    reverse = order_asc == "-1"
    portfolio_roles_sort = sorted(
        portfolio_roles,
        key=lambda portfolio_role: portfolio_role.PortfolioRole.portfolio.name,
        reverse=reverse
    )
```

File: tests/models/test_portfolio_role.py [hunk 3]
```diff
@@ -1,5 +1,7 @@
 import pytest
+from _pytest._code import ExceptionInfo
```
Comment (kylelaker-ccpo) [inline tests/models/test_portfolio_role.py:2]: Looks this was an unintentional import?

```suggestion
```

File: atat/domain/portfolio_roles.py [hunk 4]
```diff
@@ -24,6 +24,22 @@ def get(cls, portfolio_id, user_id):
 
         return portfolio_role
 
+    @classmethod
+    def get_all_by_user(cls, user_id: str = None) -> [PortfolioRole]:
```
Comment (kylelaker-ccpo) [inline atat/domain/portfolio_roles.py:28]:
```suggestion
    def get_all_by_user(cls, user_id: Optional[str] = None) -> List[PortfolioRole]:
```

This is not a valid type annotation as-written. We should go over type hints at some point. The new code will require also importing `from typing import List, Optional` if that's not already imported.

File: tests/models/test_portfolio_role.py [hunk 5]
```diff
@@ -291,3 +293,50 @@ def test_does_not_have_permission_set():
...
+
+
+def test__repr__():
+    perm_sets = PermissionSets.get_many(
+        [PermissionSets.VIEW_PORTFOLIO_FUNDING, PermissionSets.VIEW_PORTFOLIO_REPORTS]
+    )
+    port_role = PortfolioRoleFactory.create(permission_sets=perm_sets)
+    assert (
+        str(port_role).startswith("<PortfolioRole") is True
+    ), "PortfolioRole default description start with <PortfolioRole"
```
Comment (kylelaker-ccpo) [inline tests/models/test_portfolio_role.py:305]:
```suggestion
    assert (
        repr(port_role).startswith("<PortfolioRole") is True
    ), "PortfolioRole default description start with <PortfolioRole"
```

You can't guarantee that `__str__` calls `__repr__` and we don't want to require that behavior. We also should probably check that:

```
for perm_set in perm_sets:
  assert perm_set.name in repr(port_role)
```

or something like that.

File: tests/models/test_portfolio_role.py [hunk 6]
```diff
@@ -291,3 +293,50 @@ def test_does_not_have_permission_set():
...
+    user = UserFactory.create()
+
+    portfolio = PortfolioFactory.create(owner=owner)
+    user_portfolio_role = PortfolioRoles.add(user, portfolio.id)
+    owner_portfolio_role = PortfolioRoles.get(
+        portfolio_id=portfolio.id, user_id=owner.id
+    )
+
+    assert owner_portfolio_role.is_owner is True, "owner role should be owner"
+    assert user_portfolio_role.is_owner is False, "user role should not be owner"
```
Comment (kylelaker-ccpo) [inline tests/models/test_portfolio_role.py:319]:
```suggestion
    assert owner_portfolio_role.is_owner, "owner role should be owner"
    assert not user_portfolio_role.is_owner, "user role should not be owner"
```

File: tests/models/test_portfolio_role.py [hunk 7]
```diff
@@ -291,3 +293,50 @@ def test_does_not_have_permission_set():
...
+    user = UserFactory.create()
+
+    portfolio = PortfolioFactory.create(owner=owner)
+    user_portfolio_role = PortfolioRoles.add(user, portfolio.id)
+    owner_portfolio_role = PortfolioRoles.get(
+        portfolio_id=portfolio.id, user_id=owner.id
+    )
+
+    assert owner_portfolio_role.is_manager is True, "owner role should be manager"
+    assert user_portfolio_role.is_manager is False, "user role should not be manager"
```
Comment (kylelaker-ccpo) [inline tests/models/test_portfolio_role.py:333]:
```suggestion
    assert owner_portfolio_role.is_manager, "owner role should be manager"
    assert not user_portfolio_role.is_manager, "user role should not be manager"
```

**Review by kylelaker-ccpo — CHANGES_REQUESTED** [review]

**Review by jasonburkert-ccpo — CHANGES_REQUESTED** [review]
Please rename file `templates/admin/fragments/user_portafolios_roles.html` to `templates/admin/fragments/user_portfolios_roles.html`

**Comment by jasonburkert-ccpo** [comment]
> Please rename file `templates/admin/fragments/user_portafolios_roles.html` to `templates/admin/fragments/user_portfolios_roles.html`

Agree if not used, just remove.

**Comment by jGonzalez-CCPO** [comment]
@jasonburkert-ccpo https://github.com/dod-ccpo/atat/pull/2378/commits/fc0e8efea60ea3c8a0adde8d974bc7077daefd29

**Comment by jasonburkert-ccpo** [comment]
Yep, see the updates now.  Thanks.

**Comment by jGonzalez-CCPO** [comment]
wait i adding libraries

**Comment by jasonburkert-ccpo** [comment]
Yeah, seeing `script/update` failure.
```
jasonburkert@DESKTOP-FCCO10J:~/atat$ . /home/jasonburkert/atat/.venv/bin/activate
(.venv) jasonburkert@DESKTOP-FCCO10J:~/atat$ script/update
mkdir: cannot create directory ‘uploads’: File exists
Installing dependencies from lock file

No dependencies to install or update
yarn install v1.22.5
[1/4] Resolving packages...
success Already up-to-date.
Done in 0.45s.
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
Traceback (most recent call last):
  File "./script/seed_roles.py", line 10, in <module>
    from atat.app import make_app
  File "/home/jasonburkert/atat/atat/app.py", line 18, in <module>
    from atat.domain.portfolios import Portfolios
  File "/home/jasonburkert/atat/atat/domain/portfolios/__init__.py", line 1, in <module>
    from .portfolios import (
  File "/home/jasonburkert/atat/atat/domain/portfolios/portfolios.py", line 9, in <module>
    from atat.domain.invitations import PortfolioInvitations
  File "/home/jasonburkert/atat/atat/domain/invitations.py", line 6, in <module>
    from atat.domain.portfolio_roles import PortfolioRoles
  File "/home/jasonburkert/atat/atat/domain/portfolio_roles.py", line 12, in <module>
    class PortfolioRoles(object):
  File "/home/jasonburkert/atat/atat/domain/portfolio_roles.py", line 28, in PortfolioRoles
    def get_all_by_user(cls, user_id: Optional[str] = None) -> List[PortfolioRole]:
NameError: name 'Optional' is not defined
```

File: tests/models/test_portfolio_role.py (hunk 3 above)
Comment (jGonzalez-CCPO) [reply tests/models/test_portfolio_role.py]: I was testing a way to make an assertion of an exiting exception. it never works. approved your change here.

File: atat/routes/admin/admin.py (hunk 2 above)
Comment (jGonzalez-CCPO) [reply atat/routes/admin/admin.py]: super smart, I didn't think about this. more clear.

File: atat/routes/admin/admin.py (hunk 1 above)
Comment (jGonzalez-CCPO) [reply atat/routes/admin/admin.py]: make sance

File: atat/models/portfolio_role.py [hunk 8]
```diff
@@ -133,6 +133,32 @@ def can_resend_invitation(self):
...
+        return is the user of the current portfolio role is the a manager of the portfolio target on the role.
+        :return:
+        """
+        if not (self.has_permission_set and self.is_active):
+            return False
+
+        permission_sets_list = [
+            permission_set.name for permission_set in self.permission_sets
+        ]
+        return "edit_portfolio_admin" in permission_sets_list
```
[the remark this answers is not in the document]
Comment (jGonzalez-CCPO) [reply atat/models/portfolio_role.py]: nice to know about has_permission_set

File: tests/models/test_portfolio_role.py [hunk 9]
```diff
@@ -291,3 +293,50 @@ def test_does_not_have_permission_set():
...
+
+
+def test_get_all_by_user():
+    owner = UserFactory.create()
+    portfolio = PortfolioFactory.create(owner=owner)
+    get_portafolios_role_list = PortfolioRoles.get_all_by_user(user_id=owner.id)
+    assert (
+        len(get_portafolios_role_list) > 0
+    ), "a valid user id should return a list of the roles of the user."
```
[the remark this answers is not in the document]
Comment (jGonzalez-CCPO) [reply tests/models/test_portfolio_role.py]: Spanglish Typo :( - thank for catching that.

File: templates/admin/fragments/user_portafolios_roles.html [hunk 10]
```diff
@@ -0,0 +1,10 @@
+<!DOCTYPE html>
```
[the remark this answers is not in the document]
Comment (jGonzalez-CCPO) [reply templates/admin/fragments/user_portafolios_roles.html]: not need for this for now.

File: atat/domain/portfolio_roles.py (hunk 4 above)
Comment (jGonzalez-CCPO) [reply atat/domain/portfolio_roles.py]: Doing the librarires now.

**Comment by jGonzalez-CCPO** [comment]
testing changes now on local

File: tests/models/test_portfolio_role.py (hunk 5 above)
Comment (jGonzalez-CCPO) [reply tests/models/test_portfolio_role.py]: jum, good to know.

File: tests/models/test_portfolio_role.py (hunk 7 above)
Comment (jGonzalez-CCPO) [reply tests/models/test_portfolio_role.py]: ok

**Comment by jGonzalez-CCPO** [comment]
@jasonburkert-ccpo now is tested and ready for PR

File: tests/models/test_portfolio_role.py (hunk 5 above)
Comment (jasonburkert-ccpo) [reply tests/models/test_portfolio_role.py]: @kylelaker-ccpo Had to look this up.  I didn't know there were two ways to get a string representation.

Comment (jasonburkert-ccpo) [reply tests/models/test_portfolio_role.py]: Oh I see, PortfolioRole class implements __repr__ directly and includes the tested string.

**Comment by jGonzalez-CCPO** [comment]
need to fixed a style

**Review by jasonburkert-ccpo — APPROVED** [review]

**Review by jasonburkert-ccpo — APPROVED** [review]
I think this is ready to merge.
image

**Review by kylelaker-ccpo — APPROVED** [review]

**Pull request merged**